apps/nestjs-api/src/reminder-emails/reminder-emails.service.ts

```diff
@@ -97,6 +103,66 @@ export class ReminderEmailsService {
     return []
   }
 
+  async getUnmatchedMenteesFor45Days() {
+    const approvedDate = new Date()
+    approvedDate.setDate(approvedDate.getDate() - 9)
+
+    // 1st Step: Get all approved mentees from the last 7 or 14 days
+    const approvedMentees = await this.conProfilesService.findAll({
+      'RecordType.DeveloperName': UserType.MENTEE,
+      Profile_Status__c: ConnectProfileStatus.APPROVED,
+      Profile_First_Approved_At__c: {
+        $eq: jsforce.SfDate.toDateLiteral(approvedDate),
+      },
+    })
+
+    const approvedMenteeIds = approvedMentees.map(
+      (mentee) => mentee.props.userId
+    )
+
+    if (approvedMenteeIds.length === 0) return []
+
+    // 2nd Step: Get all mentorship matches where the mentee is one of the approved mentees above
+    const mentorshipMatches = await this.conMentorshipMatchesService.findAll({
+      'Mentee__r.id': {
+        $in: approvedMenteeIds,
+      },
+    })
+
+    // 3rd Step: Get all mentorship matches where the mentee is
+    // one of the approved mentees above and the staus is APPLIED
+    const appliedMentorshipMatches =
+      await this.conMentorshipMatchesService.findAll({
+        Status__c: MentorshipMatchStatus.APPLIED,
+        'Mentee__r.id': {
+          $in: approvedMenteeIds,
+        },
+      })
+
+    const mentorshipMatchMenteeIds = mentorshipMatches.map(
+      (match) => match.props.menteeId
+    )
+
+    const appliedMentorshipMatchMenteeIds = appliedMentorshipMatches.map(
+      (match) => match.props.menteeId
+    )
+
+    // 4th Step: Find approved mentees that do not have a mentorship match or waiting applied
+    const menteeIdsWithoutMentorshipMatchOrApplied = union(
+      difference(approvedMenteeIds, mentorshipMatchMenteeIds),
+      appliedMentorshipMatchMenteeIds
+    )
```

Comment:
We want to send a _Platform And New Mentors Reminder_ to mentees approved 45 days ago (or is it actually - approved >= 45 days ago?) AND who have no match AND no pending applications sent.

Do we check also mentees' pending applications here or am I confused with the naming?

Reply:
Under here I meant the `getUnmatchedMenteesFor45Days` function.

Reply:
Hey Kate, let me explain how this works:

First we find all approved mentees exactly 45 days ago. But of course some of them have already applied for mentorships during that 45 days period, so the "difference" function takes care of that. It finds the mentees that are approved BUT don't have any mentorship match yet.

The problem is, having a mentorship match doesn't mean they have a running mentorship, because fetching mentorship matches from Salesforce also returns the applications. So the "difference" unfortunately removes the mentees with active applications too. To fix that, I'm using "union" to merge the first set of mentees with the mentees that have "Applied" status mentorship matches.

Hope it's clear.

Reply:
The problem is, either it's "Status: Applied" or "Status: Accepted", we create a "MentorshipMatch" object for a Mentee/Mentor tuple, hence even in the Applied status, it's a "Mentee with Mentorship Match"

But fair, let me try an attempt to improve them 👍